src/windows.rs

```diff
@@ -33,17 +39,33 @@ pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
                 BCRYPT_USE_SYSTEM_PREFERRED_RNG,
             )
         };
-        // NTSTATUS codes use the two highest bits for severity status.
-        if ret >> 30 == 0b11 {
-            // We zeroize the highest bit, so the error code will reside
-            // inside the range designated for OS codes.
-            let code = ret ^ (1 << 31);
-            // SAFETY: the second highest bit is always equal to one,
-            // so it's impossible to get zero. Unfortunately the type
-            // system does not have a way to express this yet.
-            let code = unsafe { NonZeroU32::new_unchecked(code) };
-            return Err(Error::from(code));
+        if ret > 0 {
+            // Failed. Try RtlGenRandom as a fallback.
+            let fallback_ret = fallback_rng(chunk);
+            // NTSTATUS codes use the two highest bits for severity status.
+            if fallback_ret > 0 {
+                // We zeroize the highest bit, so the error code will reside
+                // inside the range designated for OS codes.
+                let code = ret ^ (1 << 31);
+                // SAFETY: the second highest bit is always equal to one,
```

Comment:
The code before this patch considered that warning values for NTSTATUS are not a failure (that's why it was checking for the two most significant bits of the return value from `BCryptGenRandom`). This is what this remark explains (although I think it's a mistake to consider that warning values are not a failure).

If in this patch we consider that warning values for NTSTATUS are a failure (and I think we should), then this remark does not hold anymore and should be updated. But I think we can still use `NonZeroU32::new_unchecked`, because `NTSTATUS` warning values start with 0x80000001 STATUS_GUARD_PAGE_VIOLATION, so there is no 0x80000000.

Reply:
I'm a fan of patches doing one thing at a time, so I'd prefer to restrain ourselves to not having opinions about the return values (like I accidentally did). If we wish to propose a change so warnings are failures as well (now that we have a fallback to attempt), I feel like that ought to be its own Issue and PR.

Reply:
OK, good for me then.

Reply:
I agree that our handling of error values here is not ideal. We should mimic what `NT_SUCCESS(Status)` does and only treat values `0 − 0x7FFFFFFF` as "succeeding". However, this is better addressed in a followup PR.